user_docs/en/changes.t2t

```diff
@@ -74,7 +74,19 @@ Setting ``numCells`` is still supported for single line braille displays and ``n
  - Example: ``speech.speakSsml('<speak><prosody pitch="200%">hello</prosody><break time="500ms" /><prosody rate="50%">John</prosody></speak>')``
  - The SSML parsing capabilities are backed by the ``SsmlParser`` class in the ``speechXml`` module.
  -
-- The file names of the NVDA Controller Library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+- Changes to the NVDA Controller Client library:
+ - The file names of the library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+ - Added the following functions to the controller client: (#15734, #11028, #5638, @LeonarddeR)
+  - ``nvdaController_getProcessId``: To get the process id (PID) of the current instance of NVDA the controller client is using.
+  - ``nvdaController_speakSsml``: To instruct NVDA to speak according to the given SSML. This function also supports:
+    - Providing the symbol level.
+    - Providing the priority of speech to be spoken.
+    - Speaking both synchronously (blocking) and asynchronously (instant return).
+    - The exported ``_nvdaController_onSsmlMarkReached`` pointer can be assigned to a callback function that is called in synchronous mode for every ``<mark />`` tag encountered in the SSML sequence.

```

Comment:
It is very likely my lack of understanding of how RPC rather than a real problem, but looking at this I cannot figure out how two applications wanting to provide their own callback for `onSsmlMarkReached` can do so. Since it seems to be necessary to assign the callback to the global pointer, it appears that, to be on the safe side, application developer should assign the old  value of `onSsmlMarkReached` to a temporary variable before passing any speech to NVDA, assign their own callback, speak whatever has to be spoken, and restore the old value afterwards. Would it be possible to pass pointer to `onSsmlMarkReached` as an additional parameter to `nvdaController_speakSsml` instead of using the single global callback?

Reply:
> it appears that, to be on the safe side, application developer should assign the old value of `onSsmlMarkReached` to a temporary variable before passing any speech to NVDA, assign their own callback, speak whatever has to be spoken, and restore the old value afterwards. 

That's not necessary. The old value is nullptr anyway and the internal nvdaController_onSsmlMarkReached only calls the exported pointer when it is set correctly.

> Would it be possible to pass pointer to `onSsmlMarkReached` as an additional parameter to `nvdaController_speakSsml` instead of using the single global callback?

That would be a lot nicer indeed, though I don't think we'd be able to pass the function pointer over RPC.

I'll have a look into this.

Reply:
> > it appears that, to be on the safe side, application developer should assign the old value of `onSsmlMarkReached` to a temporary variable before passing any speech to NVDA, assign their own callback, speak whatever has to be spoken, and restore the old value afterwards.
> 
> That's not necessary. The old value is nullptr anyway and the internal nvdaController_onSsmlMarkReached only calls the exported pointer when it is set correctly.


Backing up and restoring the old function pointer seems to still be necessary, without doing so I see no way in which two applications can provide their own callback. Without backing up and restoring, the last assigned callback will always be used.

Reply:
Every application can provide its own callback because the callback is stored in the user space of the client. The server, NVDA in this case, doesn't know anything about what the callback does. A callback is just RPC in the opposite direction.